packages/collections/persistent/map.pony

```diff
@@ -123,29 +143,42 @@ class val _MapNode[K: (mut.Hashable val & Equatable[K] val), V: Any val]
     let i = _Hash.array_index(nodemap, datamap, idx)
     if i == -1 then error end
     match entries(i.usize())
-    | let l: _Leaf[K, V] => l.value
+    | let l: _MapLeaf[K, V] => l.value
     | let sn: _MapNode[K, V] =>
       sn(hash, key)
     else
-      let cn = entries(i.usize()) as Array[_Leaf[K, V]] val
+      let cn = entries(i.usize()) as Array[_MapLeaf[K, V]] val
       for l in cn.values() do
         if l.key == key then return l.value end
       end
       error
     end
 
+  fun val get_positions(positions: Array[(U32, K)]): Array[(U32, K)] ? =>
```

Comment:
It seems like it should be possible to iterate over this structure without creating this new "positions" array in the process.  I believe this is similar to @sylvanc's complaint about your original implementation of this feature.  We want to avoid allocating what is effectively a copy of the entire structure when we do this iteration, so we need to find a way to move this functionality into the iterator class so that it can be performed one element at a time.